Faster parallel processing pipeline

graeme-winter opened the pull request:
WIP

Parallel processing pipeline - making a PR so work visible.

graeme-winter committed:
First cut at a start of a fast_dp with DIALS script

graeme-winter committed:
Forgot critical parameter

graeme-winter committed:
Reassign oscillation range

Critical to get the spots to match up correctly in reciprocal space; once again
shows that we have some broken data models here. Though could I have used scan
slicing for this?

graeme-winter committed:
Commentary

graeme-winter committed:
Incorporate symmetry, resolution estimation and scaling

graeme-winter committed:
Rename to fp3 -

faster parallel processing pipeline

graeme-winter committed:
First cut at tool docs

graeme-winter committed:
Rewrite

graeme-winter committed:
Derpee -> fp3

graeme-winter committed:
Logic: if image[0] != 1

graeme-winter committed:
Start factoring out

Also fix a bug where total range does not divide nicely into 5° chunks
(e.g. 0.15° images)

graeme-winter committed:
Images for indexing

graeme-winter committed:
Use index_blocks

Does involve some shuffling around between computer numbers and people
numbers

graeme-winter committed:
fp3 -> FP3 because class

graeme-winter committed:
format_phil_include: workaround

Tool to extract a subset of a phil scope which has been parsed back to text

graeme-winter committed:
Allow passing PHIL overrides

Includes scopes for every DIALS program which is used. For each step will
extract relevant PHIL parameters to a file and pass on to the program in
question if not empty.

graeme-winter committed:
Define work, duh

graeme-winter commented:
OK, think this is in a state now where it's kinda sorta ready for wider input: two outstanding issues are:

- [x] needing to usefully "sum" the integrated data back into a single scan
- [ ] running the integration blocks in parallel rather than as they are here sequential

I do however believe that this _is_ a useful starting point and is surprisingly un-gorilla-like, in that there are even a handful of tests. 

Am approaching this as a two-stage PR: part 1 is fixing up the code for squash merge as a new feature here then part 2 is grabbing those diffs across to DIALS proper for inclusion in a future release once we've shaken it down in real life.

graeme-winter committed:
New implementation of combine

Merge reflection files and experiments manually, with local implementation

graeme-winter committed:
Use local implementation of combine

graeme-winter commented:
Would particularly welcome opinions on `tools.py::combine_{experiments,reflections}`

graeme-winter committed:
Fix off-by-one; use scan slicing

graeme-winter committed:
+= boost_adaptbx.

graeme-winter committed:
dials.resolutionizer -> dials.estimate_resolution

graeme-winter committed:
Parallelism

Add future parallels wrapper to run each processing task on a given
number of cores for a given number of workers - needs cluster addition but
gives a framework for parallel operation

graeme-winter committed:
WIP: create script to perform the processing for drmaa

graeme-winter committed:
unlink

graeme-winter committed:
Run shell scripts - useful step along the way to executing on cluster

graeme-winter committed:
Look for the environment script

Anthchirp commented on fp3/util.py:
```diff
@@ -0,0 +1,78 @@
+    try:
+        return int(os.environ.get("NSLOTS"))
+    except (ValueError, TypeError):
+        pass
+
+    return number_of_processors(return_value_if_unknown=-1)
```
since we have `psutil`:
```suggestion
        return len(psutil.Process().cpu_affinity())
```
gets the number of CPUs that are available to the process, rather than the number of CPUs in the system. This is the Windows-compatible version of [`len(os.sched_getaffinity(0))`](https://docs.python.org/3.8/library/os.html#os.sched_getaffinity)

`NSLOTS` still takes priority though.

graeme-winter replied:
Computer says no:
```
>>> import psutil
>>> psutil.Process().cpu_affinity()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Process' object has no attribute 'cpu_affinity'
```

dagewa replied:
`dials.util.mp.available_cores` should handle this correctly

graeme-winter replied:
At the risk of repeating myself, this is not in a release 

```
>>> dials.util.mp.available_cores
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'dials.util.mp' has no attribute 'available_cores'
```

graeme-winter replied:
https://github.com/dials/dials_scratch/pull/5#discussion_r498694120 for reference

dagewa replied:
I did not see the earlier comment because it was hidden by being resolved.

Anthchirp replied:
This was 25 days ago. These days you should use `dials.util.mp.available_cores()` from https://github.com/dials/dials/pull/1430 instead of your `nproc()` implementation

isikhar committed:
Implement integrate task execution via drmaa array

graeme-winter committed:
A regular copy is enough, and much quicker

graeme-winter committed:
Added elementary logging

graeme-winter committed:
Show final crystal

graeme-winter committed:
New file not append mode

graeme-winter committed:
Print merging stats

graeme-winter committed:
typos

graeme-winter commented:
Usage:

```
dials_scratch.fp3 max_workers=288 parallelism=drmaa worker_nproc=4 /dls/i04/data/2020/mx27124-2/Insulin/Insulin15/Insulin15_3_master.h5
```

(for example) gives:

```
parallelism = process *drmaa
max_workers = 288
worker_nproc = 4

Finding spots...
Indexing...
Crystal:
    Unit cell: 68.0507(14), 68.0518(14), 68.0828(13), 109.5291(5), 109.5104(6), 109.5189(6)
    Space group: P 1
    U matrix:  {{-0.4845, -0.6737,  0.5580},
                {-0.8493,  0.2093, -0.4847},
                { 0.2097, -0.7088, -0.6735}}
    B matrix:  {{ 0.0147,  0.0000,  0.0000},
                { 0.0052,  0.0156,  0.0000},
                { 0.0090,  0.0090,  0.0180}}
    A = UB:    {{-0.0056, -0.0055,  0.0101},
                {-0.0158, -0.0011, -0.0087},
                {-0.0067, -0.0171, -0.0121}}
Integrating...
Combining...
Determining symmetry...
Crystal:
    Unit cell: 78.547, 78.547, 78.547, 90.000, 90.000, 90.000
    Space group: I 2 3
    U matrix:  {{ 0.8291,  0.3902,  0.4005},
                { 0.3201, -0.9185,  0.2322},
                { 0.4585, -0.0643, -0.8864}}
    B matrix:  {{ 0.0127,  0.0000,  0.0000},
                {-0.0000,  0.0127,  0.0000},
                {-0.0000, -0.0000,  0.0127}}
    A = UB:    {{ 0.0106,  0.0050,  0.0051},
                { 0.0041, -0.0117,  0.0030},
                { 0.0058, -0.0008, -0.0113}}
    A sampled at 14401 scan points
Scaling...
```

scaling stat output is still WIP...

graeme-winter committed:
Processing times

graeme-winter committed:
Correctly get scaling output

graeme-winter committed:
TODO

graeme-winter commented:
Scaling stats now gathered:

```
                                             Overall    Low     High
High resolution limit                           1.37    3.71    1.37
Low resolution limit                           55.54   55.59    1.39
Completeness                                   92.2    99.9    19.8
Multiplicity                                  100.2   128.9     4.5
I/sigma                                        49.0   205.2     0.4
Rmerge(I)                                     0.066   0.038   1.797
Rmerge(I+/-)                                  0.065   0.037   1.793
Rmeas(I)                                      0.066   0.038   2.036
Rmeas(I+/-)                                   0.066   0.037   2.051
Rpim(I)                                       0.006   0.003   0.944
Rpim(I+/-)                                    0.008   0.004   0.986
CC half                                       1.000   1.000   0.348
Anomalous completeness                         88.2    99.9     1.5
Anomalous multiplicity                         53.2    71.2     4.2
Anomalous correlation                         0.573   0.727  -0.926
Anomalous slope                               0.837
dF/F                                          0.025
dI/s(dI)                                      1.111
Total observations                          1586799  118609     777
Total unique                                  15830     920     172
```

this is ugly, but works.

graeme-winter commented:
Full run:

```
cs04r-sc-com99-01 14400-i04-ins :) $ dials_scratch.fp3 max_workers=288 parallelism=drmaa worker_nproc=4 /dls/i04/data/2020/mx27124-2/Insulin/Insulin15/Insulin15_3_master.h5
parallelism = process *drmaa
max_workers = 288
worker_nproc = 4

Finding spots...
Indexing...
Crystal:
    Unit cell: 68.0507(14), 68.0518(14), 68.0828(13), 109.5291(5), 109.5104(6), 109.5189(6)
    Space group: P 1
    U matrix:  {{-0.4845, -0.6737,  0.5580},
                {-0.8493,  0.2093, -0.4847},
                { 0.2097, -0.7088, -0.6735}}
    B matrix:  {{ 0.0147,  0.0000,  0.0000},
                { 0.0052,  0.0156,  0.0000},
                { 0.0090,  0.0090,  0.0180}}
    A = UB:    {{-0.0056, -0.0055,  0.0101},
                {-0.0158, -0.0011, -0.0087},
                {-0.0067, -0.0171, -0.0121}}
Integrating...
Combining...
Determining symmetry...
Crystal:
    Unit cell: 78.547, 78.547, 78.547, 90.000, 90.000, 90.000
    Space group: I 2 3
    U matrix:  {{ 0.8291,  0.3902,  0.4005},
                { 0.3201, -0.9185,  0.2322},
                { 0.4585, -0.0643, -0.8864}}
    B matrix:  {{ 0.0127,  0.0000,  0.0000},
                {-0.0000,  0.0127,  0.0000},
                {-0.0000, -0.0000,  0.0127}}
    A = UB:    {{ 0.0106,  0.0050,  0.0051},
                { 0.0041, -0.0117,  0.0030},
                { 0.0058, -0.0008, -0.0113}}
    A sampled at 14401 scan points
Scaling...
Estimating resolution...
Recommendation is use all data
                                             Overall    Low     High
High resolution limit                           1.37    3.71    1.37
Low resolution limit                           55.54   55.59    1.39
Completeness                                   92.2    99.9    19.8
Multiplicity                                  100.2   128.9     4.5
I/sigma                                        49.0   205.2     0.4
Rmerge(I)                                     0.066   0.038   1.797
Rmerge(I+/-)                                  0.065   0.037   1.793
Rmeas(I)                                      0.066   0.038   2.036
Rmeas(I+/-)                                   0.066   0.037   2.051
Rpim(I)                                       0.006   0.003   0.944
Rpim(I+/-)                                    0.008   0.004   0.986
CC half                                       1.000   1.000   0.348
Anomalous completeness                         88.2    99.9     1.5
Anomalous multiplicity                         53.2    71.2     4.2
Anomalous correlation                         0.573   0.727  -0.926
Anomalous slope                               0.837
dF/F                                          0.025
dI/s(dI)                                      1.111
Total observations                          1586799  118609     777
Total unique                                  15830     920     172
Total processing time: 754.60s
```

graeme-winter commented:
https://github.com/dials/dials/issues/1438 would enable more flexible reporting of output from individual steps...

graeme-winter commented:
For reference; output of xia2 with the same data:

```
Environment configuration...
Python => /dls/science/users/gw56/cctbx/build/../conda_base/bin/python3.6
CCTBX => /dls/science/users/gw56/cctbx/modules
CCP4 => /dls_sw/apps/ccp4/7.1.003/ccp4-7.1
CCP4_SCR => /tmp/1259482.1.test-high.q/tmp3d630snq
Starting directory: /dls/mx-scratch/gw56/fp3/14400-i04-ins-xia2-ref
Working directory: /dls/mx-scratch/gw56/fp3/14400-i04-ins-xia2-ref
Free space:        90817.33 GB
Host: cs04r-sc-com99-07
Contact: xia2.support@gmail.com
XIA2 0.7.92-gd5332161
DIALS 3.dev.210-ge875b210c
CCP4 7.1.003
Command line: xia2 image=/dls/i04/data/2020/mx27124-2/Insulin/Insulin15/Insulin15_3_master.h5
Project directory: /dls/mx-scratch/gw56/fp3/14400-i04-ins-xia2-ref
-------------------- Spotfinding SWEEP1 --------------------
547447 spots found on 14400 images (max 10616 / bin)
           **             **             **             **  
   ****    ***   *****    ***    ****    ***   *****    *** 
*************************************** ********************
************************************************************
************************************************************
************************************************************
************************************************************
************************************************************
************************************************************
************************************************************
1                        image                         14400
------------------- Autoindexing SWEEP1 --------------------
All possible indexing solutions:
cI  78.63  78.63  78.63  90.00  90.00  90.00
hR 111.19 111.19  67.90  90.00  90.00 120.00
tI  78.63  78.63  78.63  90.00  90.00  90.00
oI  78.63  78.63  78.63  90.00  90.00  90.00
oF  78.55 110.99 111.23  90.00  90.00  90.00
mC  78.53 111.15  67.89  90.00 125.17  90.00
aP  67.89  68.05  68.04 109.52 109.43 109.40
Indexing solution:
cI  78.63  78.63  78.63  90.00  90.00  90.00
-------------------- Integrating SWEEP1 --------------------
Processed batches 2 to 14401
Standard Deviation in pixel range: 0.52 1.22
Integration status per image (60/record):
%ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooo%oooooooooooo%oooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooo%oooooooooooooooooooooooooooooooooo%ooooo%ooooooooo
oo%ooo%oo%oo%ooooooooooooooooooooo%ooo%o%ooooo%oo%o%oooooooo
%%%ooo%oo%o%ooooooooo%oo%ooo%%%oo%%oooo%oo%oooooooo%o%%ooo%o
o%oooo%ooooo%o%o%%%%o%%ooo%o%oooooooooo%%ooooo%o%oo%o%oooooo
%%%%oo%%%ooooooooooo%o%oo%%%oooooooo%%%o%o%o%oooo%o%oo%ooo%o
o%%%oooo%o%oo%o%%oo%oo%o%%%o%%%%%%oo%%%o%oooo%%%o%%o%%o%%oo%
%o%%%%%ooo%%oo%o%%ooo%%oooo%oo%oo%%%o%oooo%%%%oo%%%o%%%oooo%
%%%%%%%o%oo%oo%%%%ooooo%%o%oooo%%o%o%o%o%o%o%%%o%o%%%%oo%ooo
o%%oooo%%%o%oo%%%o%o%%%%%ooo%oooo%oo%%o%o%oo%oooo%%%oo%%%%%o
o%%o%%o%%%ooo%%%o%%%oo%o%ooooooo%%o%%%oo%%oo%%o%o%%%o%%oo%%o
oooo%o%o%ooo%%oooo%%ooooooo%oooo%oo%o%%%%oooooo%%%%%ooo%%%o%
o%oo%oooooo%o%%%oo%ooo%oo%%oooooo%ooo%%ooooooooooo%ooo%oooo%
oooooo%o%oo%oo%o%ooooooo%oo%ooo%ooo%ooooooooo%%ooooooo%o%%%o
ooo%%ooooooooo%o%oo%o%ooooooooooo%oooo%ooooooooooooooooooooo
o%ooooooooooooooooooooooooooooooooo%ooooooooooo%%ooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooo%ooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooo%oooooooooooooooooooooooooooooooo%oooooo%oooo
ooo%oooooooooooooooooooooooooooooo%ooo%oooooooo%oooooooooooo
ooo%oooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooo%oooo%%%oooooooo%ooooo%oooooo%oooooooooo%oooooo%oooooo%o
o%oooo%oooooo%ooo%o%%%o%%o%%o%oooo%oooo%ooooo%%o%ooooo%oo%%o
oo%ooooooo%oo%o%ooooooo%%%ooo%%o%%ooooo%o%ooo%%%oo%%oo%ooo%o
ooo%%%o%oo%%o%%%%%%%oo%%%%oo%oo%o%%oooo%%%%%%%%ooooo%%o%%o%%
oo%%o%%o%%o%%%%ooo%%o%o%ooo%%%%%%o%o%%%%%%%o%o%%%oo%o%%%%%%%
o%o%%o%o%%%oo%%o%%%%%%%%%%%oo%%%%o%%%%%%oooo%o%%%%%%o%o%o%%%
%o%%%%oo%%%%oo%o%%%%%%%%oo%oo%%o%%%%oo%%%oo%o%%%%%%%%o%%%%o%
%%%o%%%%%o%%%%%%%%o%%o%%%%%%o%%%%%oo%%%%%o%%%%%%o%%%o%o%o%%%
o%%%%o%%%%o%%%%%%%%%%o%%ooo%ooo%%%oooo%o%%%%oo%o%%o%%o%o%%%o
%%oo%o%%oo%ooo%o%%%o%%oo%o%%%ooooo%%o%oo%%ooo%oooo%%oo%oo%%o
ooo%oooo%o%oooooooo%ooooo%o%ooooooo%%ooooooooooooooooooooooo
ooooooooooooooooo%o%oooooooooooo%ooooooooooooooooooooooooooo
ooooooo%ooooooooo%oooo%oo%oooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooo%ooooooooooooooooooooooooooo%oooooooo%ooooooooo
oooooooooooooooooooooo%o%%oooooooooooooooo%oooooooooooooooo%
%o%%oo%oo%ooooooooo%ooooooooo%oooooooo%%ooooo%ooo%%oo%%ooooo
o%oo%o%oo%ooo%oooooooooo%o%oo%%oo%%ooooooo%o%%o%ooooo%oooo%o
o%%o%oo%oo%o%ooooo%ooo%ooo%%oooo%o%ooooo%oo%%oooooo%%%oo%ooo
o%%%o%%o%%oooooo%ooo%oooo%oo%%%oo%o%oo%o%%%o%o%oo%ooo%%o%o%%
oo%%o%o%oo%oo%o%%oo%oo%%%oooo%o%%%%oo%o%%oooo%%%o%%o%ooo%o%%
%%%%o%%oooo%%o%%%%%o%%%%oooooooo%%%o%%ooooo%%%o%o%ooo%%ooo%%
%%%%%%%%%o%%%o%%%%%%%o%%o%o%ooo%%%%o%o%%%oo%o%%o%o%%%%%o%%o%
o%%%o%ooo%o%o%%%%o%o%o%%%oooo%oooo%oo%%o%%o%o%ooo%%%%%o%%%oo
oo%%%%%o%%oo%o%%%o%%ooo%%%oo%oo%%%o%o%%o%oo%o%oooo%ooo%%o%oo
ooooooo%%ooo%oo%ooo%oooooo%ooooooo%o%%oo%oooo%%%%oooooo%o%o%
oooo%oo%o%o%oo%ooo%oooo%oooooooooo%o%%%ooooooooo%o%ooo%oooo%
o%ooooo%oo%ooooo%oooo%oooo%%oooooooooooooooooooooooo%o%ooooo
%ooooooooo%oooooooo%%ooooo%%oooooooooo%oooooooooooooooo%oooo
ooo%ooooooooo%%oooo%oooooooooooooooooo%ooooooooo%ooooooooooo
oooooooooooooooooooooooooooooooooooooooooooo%ooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooo%oooooooooooooo%oooooooooooooooooooooooooooooooooooooooo
oooooooooooo%oooooooooooooooooo%ooooooooooooooooooo%oooooo%o
oo%oooo%o%o%ooooooooooooo%%ooooooo%ooo%oo%oo%%o%ooooo%ooo%o%
%oooo%o%oooo%oooo%oo%o%%%ooooooooo%oo%oooooooo%ooooo%o%%%%%o
%ooooo%ooo%ooo%%oooo%o%oo%oo%%o%o%ooooo%ooooo%o%ooo%%%%o%ooo
ooo%%o%oooo%o%%%o%%o%o%%%ooo%o%%o%%%%oo%%%o%o%%ooo%%%o%ooo%%
oo%oo%%ooo%ooo%oo%%%%%o%%o%o%%oo%ooo%%%%%%%%%%%%%o%%%%o%%%%o
o%%o%%oo%%o%o%o%%%%%%%%%%o%oo%%%%%%%o%%%o%oo%o%%%%%%%ooo%%%o
%%o%%%oo%%%%%o%%%%%%o%%oo%o%o%%oo%o%o%o%%oo%o%o%%o%%oo%%%o%%
%%%o%%%%%%%%%o%%%%o%%oo%%%o%oo%%%%oo%%%%%o%%%%%%%%o%%%o%%o%%
%%%%%o%o%ooo%o%%%%%%%%%%%%oo%%%o%%%o%o%oo%o%%%%%%oo%%%o%%%%%
%%%%%%o%%o%%oo%%oooo%ooooo%oooo%%%o%o%%o%oooo%%oooo%%o%oo%%o
o%oooooo%%%%o%ooo%ooooo%o%o%%%oooooooooo%ooo%o%ooo%ooooooooo
oooooooo%ooooooooooooo%ooooooooo%ooo%oooooooooooooooooo%oooo
ooooooooooooooooooooooooooooooooooooooooooooooooooooo%oooooo
%ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooo%ooooooo
oooooooooo%oooooooooo%ooooooooo%oooooooo%oooooooooooo%oooooo
%oooooooo%ooooooooooooooooooo%oooo%oooooooooooooo%oo%ooooooo
ooooo%ooooo%ooo%%oooo%oo%o%ooo%oo%ooooooooooo%oo%oooo%oo%o%o
oo%o%%oooooo%oooo%%%o%oo%ooo%oooooooooooo%%o%ooooooo%o%o%ooo
o%%ooooo%%oo%oo%%%oo%oooo%oo%ooooo%%oo%ooo%oo%ooooo%%o%o%o%o
o%%%%o%ooo%%o%o%%o%oooo%o%o%%o%%%%oo%%o%o%ooooo%oo%oooo%ooo%
%%%%o%%o%o%%o%oo%o%oo%%%oooo%%%o%o%o%ooo%ooo%%%o%%%oo%%%oo%%
%%%%%%%%%%o%%%o%%o%%%o%%ooo%oo%%%%o%%%%%%o%oo%%o%%%%%%o%%ooo
%%%%%%oo%oo%o%%%%%%o%o%%%ooo%%ooo%ooo%%oo%ooo%o%o%%%%oo%%%o%
oo%%%%%o%%ooooo%%%ooooooooo%ooo%%%o%%%%oo%oo%oooo%%%oo%o%%o%
oooo%o%o%ooo%ooooo%%o%oooo%%o%ooooo%ooo%%o%o%oo%%oo%ooooo%o%
oooo%o%ooo%%o%%%%o%oooooo%%oooooo%oooo%ooooooooooo%o%oo%ooo%
%%oooo%%o%oooooo%o%oooo%o%%%ooo%o%%%ooooooooooo%o%oooo%ooooo
ooo%%oooo%o%o%ooo%ooo%oo%%oooooooooooo%ooooooooooooooooooo%o
oo%ooooooooooooooooooooooooooooooooo%ooooooooooo%ooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooo%%oooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooo%oo%oo%%ooooo%o%oooo%%ooo%oo%o%ooooo%%%ooo%ooooooooo%oo
ooo%oo%ooo%ooo%ooo%o%%%o%ooooooooo%oo%ooooo%oo%ooooooo%oo%%o
ooooooooooo%oo%%%ooo%%ooo%%o%%%oo%ooooo%oo%o%%o%%%%oo%%ooooo
o%ooo%%o%oo%o%%%%%o%%%%%%oo%%o%%o%%o%o%%%%%o%o%ooooo%%o%%%o%
%oo%%%%o%%oooo%oo%o%%%oo%oo%%%%ooo%oooo%%%%%%%o%o%o%%o%%%%o%
oo%%%o%o%oo%%%%%%%%%%%%%o%%oo%%%%%%%o%%oo%oo%%o%%%%%%oooo%%%
%oo%%%oo%%%%%oo%%%%%%ooo%%%%%%%o%%%%o%%o%oo%o%%%%ooo%%o%%o%%
%%%%%%%%o%%%%%%%%o%%%oo%%%o%%o%o%ooo%%%%%%%%%%%%%%%%%%o%oo%%
o%%%%oo%ooo%%%%%%%%%o%%%%ooo%%o%%o%ooo%%%%o%%%%%oo%o%%ooo%oo
%%%o%%o%ooo%%o%o%o%o%o%oooo%ooooo%ooo%oo%oooo%ooo%oooo%ooooo
ooooo%%o%ooooooooo%ooooooooo%%oooooo%oooo%oooooooo%%ooooo%oo
oooooooo%ooo%oooooo%oooo%ooooooooooooo%oooooooooooooooo%oooo
ooooooo%o%ooooooo%ooooooo%%ooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo%
oooooooooo%ooooooooooooooooooo%ooooooooooooooooooooooooo%ooo
%ooooooooooooooooooooooooooo%oooooo%ooo%oooooo%oooooooo%%ooo
ooooooooo%ooooo%o%ooooooooo%oooooo%%ooo%ooooo%%oooo%oooo%o%o
ooo%%oo%o%ooooo%oo%%ooo%%o%o%ooooooo%%oo%o%ooooooooo%o%o%ooo
oo%%%o%ooooo%ooo%%%o%o%oo%oo%o%%ooo%o%%oo%o%%o%ooo%o%o%oo%o%
o%o%o%%%%oooo%o%%%%%%oooo%oo%oooo%%%o%%oo%%ooo%oo%%%%%oo%oo%
%%%%o%o%ooo%%oo%%o%%oo%%o%%oooo%%o%ooo%%oooo%%o%%%o%o%o%o%o%
o%%%%%%%%oo%o%%%%o%o%oo%ooo%ooo%%%oooo%%%oo%o%%%%%o%%%%o%%o%
%%%%%o%ooooo%o%o%o%o%o%%%%ooo%oo%%%oo%%oo%%o%%%oo%%o%oo%o%oo
ooo%o%oo%oooo%o%o%oo%oooooooooo%%%o%o%%o%oo%ooo%ooo%o%ooo%%o
%%ooooooooooo%ooo%o%o%o%o%o%o%oo%o%o%o%%%ooo%%%o%oo%o%ooo%o%
oooooo%%%oo%oo%o%o%oooo%o%ooooo%oo%%o%%oooo%oooooo%o%oo%oooo
oooooo%%o%o%oo%%%oooo%o%o%o%ooo%oo%oooooooooooo%o%oo%ooooooo
ooo%%oooo%o%ooo%o%ooo%%%o%oooo%oooo%oo%ooooooooooo%oooooo%oo
oooooooooooooooooooooooooooooooooo%%oooooooooooo%ooooooooooo
oooo%ooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooo%ooooooooooooooo
oooo%oooooooooooooo%ooooooooooooooooooooooooooooooooooooo%oo
ooooo%ooooooooooooooooooooooooo%%oo%ooooooooo%ooooo%oo%o%ooo
ooooooo%%oooooooooooo%ooooo%oooo%%ooooooo%ooooo%oooooooo%ooo
ooo%oooo%oooooooooooo%o%%%oo%%oooo%%ooooooo%oo%oo%o%%oo%o%%%
%o%%oooo%o%%ooo%oooo%oo%o%%o%%%oo%o%ooooooooooo%oo%%ooo%oo%o
o%oo%%o%ooo%%%%%%%%o%o%o%oo%ooooo%%%ooo%%%o%oo%%%oo%%%o%%o%%
o%o%oo%o%%o%%o%%%oo%%ooo%%%%%%%o%%%%o%%%%o%%o%o%%%%o%%%%%%%%
oo%oo%o%%%o%o%%%%%o%o%oo%%%%o%%%%%%%%%%%o%%o%%%%%%%%%ooo%o%%
%oo%%%%oo%o%%oo%o%%%%o%%%%%oo%%o%o%o%%%%%%o%%oo%%%o%%%%%%o%%
%%%%o%%%%o%%%%%%%%%%%%o%o%oo%o%o%o%o%oo%%%o%%%%%%o%%%%o%%oo%
o%%%%o%%%%%o%oo%%%o%o%%%%oo%%ooo%%oo%%%%%%ooo%%%oo%o%%ooo%%o
%%%ooo%oooooooo%%%oo%o%oo%oo%oooooo%o%o%%oooo%%ooo%o%o%oo%%o
oooooooooooooooooooo%oooo%%o%ooooooo%%%oooooooooo%o%oooooooo
oooooooooo%oooooooooooooooo%oooo%%oo%ooooooooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooo%ooooooooooooooooo
oooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooooooooooo%oo
"o" => good        "%" => ok        "!" => bad rmsd
"O" => overloaded  "#" => many bad  "." => weak
"@" => abandoned
Mosaic spread: 0.097 < 0.097 < 0.097
-------------------- Preparing DEFAULT ---------------------
--------------------- Scaling DEFAULT ----------------------
Resolution for sweep NATIVE/SWEEP1: 0.00
--------------------- Scaling DEFAULT ----------------------
---------------- Systematic absences check -----------------
Most likely space group: I 2 3
------------------- Unit cell refinement -------------------
Overall:  78.60  78.60  78.60  90.00  90.00  90.00
Project: AUTOMATIC
Crystal: DEFAULT
Sequence: 
Wavelength name: NATIVE
Wavelength 1.23985
Sweeps:
SWEEP SWEEP1 [WAVELENGTH NATIVE]
TEMPLATE Insulin15_3_master.h5
DIRECTORY /dls/i04/data/2020/mx27124-2/Insulin/Insulin15
IMAGES (USER) 1 to 14400
MTZ file: /dls/mx-scratch/gw56/fp3/14400-i04-ins-xia2-ref/DEFAULT/NATIVE/SWEEP1/integrate/12_integrated.refl
For AUTOMATIC/DEFAULT/NATIVE                 Overall    Low     High
High resolution limit                           1.37    3.72    1.37
Low resolution limit                           55.58   55.62    1.39
Completeness                                   92.7    99.8    22.6
Multiplicity                                  113.1   148.9     4.7
I/sigma                                        55.3   235.9     0.4
Rmerge(I)                                     0.063   0.036   2.082
Rmerge(I+/-)                                  0.062   0.035   2.076
Rmeas(I)                                      0.063   0.036   2.349
Rmeas(I+/-)                                   0.063   0.035   2.362
Rpim(I)                                       0.005   0.003   1.073
Rpim(I+/-)                                    0.007   0.004   1.117
CC half                                       1.000   1.000   0.351
Wilson B factor                              18.620
Anomalous completeness                         88.6    99.9     1.6
Anomalous multiplicity                         60.0    82.2     4.4
Anomalous correlation                         0.632   0.766  -0.628
Anomalous slope                               0.886
dF/F                                          0.026
dI/s(dI)                                      1.159
Total observations                          1796144  136816     912
Total unique                                  15886     919     194
Assuming spacegroup: I 2 3
Unit cell (with estimated std devs):
78.59555(9) 78.59555(9) 78.59555(9)
90.0        90.0        90.0       
mtz_unmerged format:
Scaled reflections (NATIVE): /dls/mx-scratch/gw56/fp3/14400-i04-ins-xia2-ref/DataFiles/AUTOMATIC_DEFAULT_scaled_unmerged.mtz
mtz format:
Scaled reflections: /dls/mx-scratch/gw56/fp3/14400-i04-ins-xia2-ref/DataFiles/AUTOMATIC_DEFAULT_free.mtz
Processing took 01h 32m 21s
XIA2 used... ccp4, dials, dials.scale, xia2
Here are the appropriate citations (BIBTeX in xia2-citations.bib.)
Beilsten-Edmands, J. et al. (2020) Acta Cryst. D76.
Winn, M. D. et al. (2011) Acta Cryst. D67, 235-242.
Winter, G. (2010) J. Appl. Cryst. 43, 186-190.
Winter, G. et al. (2018) Acta Cryst. D74, 85-97.
Status: normal termination
```

N.B. processing took 92 minutes compared with ~ 12.6 minutes for fp3 - data from `xia2` were slightly better (as one may hope) however need to do some analysis on this to see if it is just the effect of splitting the reflections across 5° blocks. Certainly results from fp3 above seem reasonable here. `xia2` job was run with atom not set...

graeme-winter committed:
Initial documentation

graeme-winter committed:
+= authors

graeme-winter committed:
Fix indentation

graeme-winter committed:
Fix indentation

graeme-winter committed:
Fix indentation

graeme-winter committed:
italic

graeme-winter committed:
If debug; print logs

graeme-winter commented:
#info some work needed on this before it will support multi-sweep i19 data sets (not a lot; could even be ~ 10 lines of Python)

dagewa commented:
Just trying a dumb run (not setting any options) just to see what it does. It looks like it doesn't like having a previous aborted run around?
```
fcx32934@zdar:/tmp$ time dials_scratch.fp3 ~/sw/cctbx/build/dials_data/thaumatin_i04/th_8_2_0*cbf

Traceback (most recent call last):
  File "/home/fcx32934/sw/cctbx/build/../modules/dials_scratch/command_line/fp3.py", line 563, in <module>
    fp3.index()
  File "/home/fcx32934/sw/cctbx/build/../modules/dials_scratch/command_line/fp3.py", line 111, in index
    indexed = ExperimentList.from_file(os.path.join(work, "indexed.expt"))
  File "/home/fcx32934/sw/cctbx/modules/dxtbx/model/__init__.py", line 861, in from_file
    return ExperimentListFactory.from_serialized_format(filename, check_format)
  File "/home/fcx32934/sw/cctbx/modules/dxtbx/model/experiment_list.py", line 739, in from_serialized_format
    return ExperimentListFactory.from_json_file(filename, check_format)
  File "/home/fcx32934/sw/cctbx/modules/dxtbx/model/experiment_list.py", line 700, in from_json_file
    with open(filename, "r") as infile:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/index/indexed.expt'

```

dagewa commented on fp3/tools.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def combine_experiments(fins, fout):
+    """Combine experiment files from find to a single output file fout. 
```
```suggestion
    """Combine experiment files from fins to a single output file fout. 
```

dagewa commented on fp3/util.py:
```diff
@@ -0,0 +1,78 @@
+
+
+def even_blocks(n0, n1, m):
+    """Split the range(n0, n1) into m evenly sized chunks, yeild start, end
```
```suggestion
    """Split the range(n0, n1) into m evenly sized chunks, yield start, end
```

dagewa commented on fp3/util.py:
```diff
@@ -0,0 +1,78 @@
+        return diff
+
+
+def nproc():
```
This could use `dials.util.mp.available_cores` now

graeme-winter replied:
Could, but if will be separate package would probably be sane to keep from needing latest features... & currently "works"

dagewa commented on command_line/fp3.py:
```diff
@@ -0,0 +1,578 @@
+            indexed = ExperimentList.from_file(os.path.join(work, "indexed.expt"))
+
+            self._experiment[0].crystal = indexed[0].crystal
+            logger.info("Picked up pre-exising crystal:")
```
```suggestion
            logger.info("Picked up pre-existing crystal:")
```

dagewa commented on command_line/fp3.py:
```diff
@@ -0,0 +1,578 @@
+        just reload the results from the previous run."""
+
+        work = os.path.join(self._root, "index")
+        if os.path.exists(work):
```
```suggestion
        if os.path.exists(os.path.join(work, "indexed.expt")):
```
to avoid stacktrace if `index/` exists, but `indexed.expt` was never written to

dagewa commented on command_line/fp3.py:
```diff
@@ -0,0 +1,578 @@
+        """Integration of the complete scan: will split the data into 5 deg
+        chunks and spin the integration of each chunk off separately"""
+
+        rng = self._rng
```
the naming here is a bit terse. I was thinking "random number generator", not "image range"

dagewa commented on command_line/fp3.py:
```diff
@@ -0,0 +1,578 @@
+}
+parallelism = *process drmaa
+  .type = choice
+max_workers = 1
+  .type = int
+  .help = "Maximum number of concurrent workers"
+worker_nproc = 1
+  .type = int
+  .help = "Number of processes for each worker"
```
It would be neat if the default was some kind of appropriate automatic selection for the local machine this is running on

graeme-winter replied:
Added in b219fc050e14e781f74c76f8b33cdb5f80ac643a

It is possibly a bit of a baroque solution but should do something sensible.

dagewa commented on command_line/fp3.py:
```diff
@@ -0,0 +1,578 @@
+        logging.basicConfig(
+            filename="dials.fp3.log", filemode="w", format="%(message)s"
+        )
+        logger.addHandler(logging.StreamHandler(sys.stdout))
```
It looks like the various logging things here are doing stuff that `dials.util.log.config` handles. It might be better to use that directly, and hence get the `ColoredFormatter` etc?

graeme-winter replied:
Could do, will leave in the queue as not familiar with `dials.util.log.config`

Is in 3.1 release series though so not against this.

dagewa reviewed (approved):
Looks useful and is certainly fast. If processing in 5 deg blocks is it even necessary to do scan-varying refinement? It could have a static model within the block - which should be robust and would be no worse than what XDS does.

graeme-winter committed:
Typo

Co-authored-by: David Waterman <dagewa@users.noreply.github.com>

graeme-winter committed:
Typo

Co-authored-by: David Waterman <dagewa@users.noreply.github.com>

graeme-winter committed:
Typo

Co-authored-by: David Waterman <dagewa@users.noreply.github.com>

graeme-winter committed:
Check file present

Co-authored-by: David Waterman <dagewa@users.noreply.github.com>

graeme-winter committed:
Disambiguate rng

graeme-winter committed:
Prime factor code for sharing out cores

graeme-winter committed:
If nproc and workers == 1 guess sensible division

Anthchirp commented:
Sorry, was not clear to me that fp3 is already set against a DIALS release, given that it’s in a PR into a scratch space and not laid out as a package yet.

graeme-winter commented:
> Sorry, was not clear to me that fp3 is already set against a DIALS release, given that it’s in a PR into a scratch space and not laid out as a package yet.

Evidently it is not yet a package etc. however we did discuss the other day making it one, so I was kinda trying to avoid picking up too much baggage from master. If we are copying the new bits over to 3.1 series then valid, I guess.

graeme-winter committed:
Cope with data sets > 5000°

graeme-winter committed:
Add to repo

graeme-winter committed:
Revert "Add to repo"

This reverts commit 299896cfca761be78ba0141a389fda12fe6a5617.

Meant to put on master :-/

graeme-winter committed:
Name it: dials.fp3

graeme-winter committed:
Procrunner API change

graeme-winter committed:
More than one experiment

graeme-winter committed:
Multi-sweeps...

Add a check at the start that only one sweep has been passed as input and
a collection of FIXME for coping with > 1 sweep / lattice in the longer
term

graeme-winter committed:
Object files -> ignore

graeme-winter committed:
Revert "Object files -> ignore"

This reverts commit 23d3231414771c15b2db39fc25a0ca0de1c05246.

graeme-winter committed:
WIP: combine partials

graeme-winter committed:
WIP continued: now more sensible

graeme-winter committed:
Remove old implementation

graeme-winter commented:
TODO: modify to use DIALS 3.2 feature set now released; reformat as package

@jbeilstenedmands  I have asked for your review for "combine_reflections" in `fp3/tools.py`

@Anthchirp would welcome input on how to close this off as a Python package now I guess... I think it is close to OK for a 0th order implementation (though obviously scope for future improvement)

graeme-winter commented:
TODO: remove duff reflections before manipulations; use vector expressions for calculating weighted sums

Anthchirp commented:
do you want to keep the development history in $newpackage?

graeme-winter commented:
> do you want to keep the development history in $newpackage?

Nope, not a chance. Without a decent rewrite the history is close to:

![git_commit](https://user-images.githubusercontent.com/1241716/98003910-393ecd00-1de7-11eb-8cb7-f13d780885c4.png)

jbeilstenedmands commented:
I tried out fp3 on the beta lactamase dataset, however I got an assertion error as part of the combine step:
`assert d0["miller_index"][i] == d1["miller_index"][j]`
Printing out the details of the assertion fail and allowing to continue, it seems that there are cases where the spots that are split across a processing boundary can be misindexed. In this case they are off by 1 in `l` for some low values of `l`, the spot positions indicate this is the same spot split across a processing boundary:
```
Mismatched indices: (-1, 26, 1) (-1, 26, 2)  Spot positions(px): (2113.23,1888.61,159.50), (2113.77,1889.74,160.50)
Mismatched indices: (15, -18, -1) (15, -18, -2)  Spot positions(px): (1197.50,424.83,269.50), (1196.50,420.00,270.50)
Mismatched indices: (19, -13, -1) (19, -13, -2)  Spot positions(px): (1460.98,453.74,269.50), (1460.77,454.77,270.50)
Mismatched indices: (15, -25, 2) (15, -25, 1)  Spot positions(px): (945.68,112.66,289.50), (945.86,112.69,290.50)
Mismatched indices: (7, -29, -3) (7, -29, -4)  Spot positions(px): (464.87,214.76,289.50), (462.00,212.50,290.50)
Mismatched indices: (9, -31, -1) (9, -31, -2)  Spot positions(px): (381.01,3.73,299.50), (380.17,3.60,300.50)
Mismatched indices: (10, -30, 1) (10, -30, 0)  Spot positions(px): (501.12,28.31,299.50), (500.00,28.86,300.50)
Mismatched indices: (7, -33, 1) (7, -33, 0)  Spot positions(px): (99.94,2522.17,519.50), (97.36,2521.79,520.50)
Mismatched indices: (22, -18, -2) (22, -18, -3)  Spot positions(px): (1443.29,2312.19,539.50), (1441.99,2312.68,540.50)
Mismatched indices: (25, -10, 0) (25, -10, -1)  Spot positions(px): (1809.73,2122.88,539.50), (1808.48,2122.59,540.50)
Mismatched indices: (19, -21, -2) (19, -21, -3)  Spot positions(px): (1233.82,2312.61,539.50), (1233.33,2313.42,540.50)
Mismatched indices: (13, -22, 4) (13, -22, 3)  Spot positions(px): (1018.01,2122.39,559.50), (1016.85,2121.21,560.50)
Mismatched indices: (18, -22, 6) (18, -22, 5)  Spot positions(px): (1175.98,2334.86,559.50), (1176.00,2333.00,560.50)
Mismatched indices: (-16, 23, 1) (-16, 23, 2)  Spot positions(px): (1472.50,165.90,639.50), (1472.71,165.03,640.50)
Mismatched indices: (-24, 17, -2) (-24, 17, -1)  Spot positions(px): (978.33,89.52,649.50), (979.07,89.07,650.50)
```
Not sure at the moment how to address this but I'll have a think.

graeme-winter commented:
@jbeilstenedmands excellent spot thank you - this needs some attention upstream of this then

Ah, wait - this is in the matching process, so what that means (I guess) is we need a smarter way of matching reflections which are found in both halves of the data set - if I don't do this (as was before) you end up with two "half-observations" which mess things up...

jbeilstenedmands commented:
I think the matching is correct, _i.e._ you have found the two halves of same spot, but one is misindexed, due to the difficulty in indexing along a certain direction for a thin slice of data? So does there need to be an extra step in this case to determine which index is correct and use that for both halves?

graeme-winter commented:
> I think the matching is correct, _i.e._ you have found the two halves of same spot, but one is misindexed, due to the difficulty in indexing along a certain direction for a thin slice of data? So does there need to be an extra step in this case to determine which index is correct and use that for both halves?

Looking at these now and remembered that this data was artificially processed to give it 0.5° images which could have some _interesting_ effects...

graeme-winter commented:
This fails going into block 16 (i.e. the 17th block) however the orientation matrix does not seem that different? 

```
Grey-Area fp3 :( $ dials.compare_orientation_matrices integrate0015/indexed.expt integrate0016/indexed.expt 
The following parameters have been modified:

input {
  experiments = integrate0015/indexed.expt
  experiments = integrate0016/indexed.expt
}

Change of basis op: a,b,c
Rotation matrix to transform crystal 1 to crystal 2:
{{1.000, 0.001, -0.000},
 {-0.001, 1.000, 0.000},
 {0.000, -0.000, 1.000}}
Rotation of -0.033 degrees about axis (0.127, 0.299, 0.946)
```

jbeilstenedmands commented on fp3/tools.py:
```diff
@@ -0,0 +1,101 @@
+    Makes assumptions that the data are consistent from one to the next, no
+    particular tests are performed at the moment."""
+
+    d0 = flex.reflection_table.from_file(fins[0])
+
+    for f in fins[1:]:
+        d1 = flex.reflection_table.from_file(f)
+
```
Testing on the x4-wide dataset, I also got a failure due to dividing by zero variance at L61. If I remember correctly, you can get a zero variance set if the reflections were not integrated.
I think all that's needed to clean the data is this:


```suggestion
    d0 = flex.reflection_table.from_file(fins[0])
    if "intensity.prf.variance" in d0:
        d0 = d0.select(d0["intensity.prf.variance"] > 0)
    d0 = d0.select(d0["intensity.sum.variance"] > 0)

    for f in fins[1:]:
        d1 = flex.reflection_table.from_file(f)
        if "intensity.prf.variance" in d1:
            d1 = d1.select(d1["intensity.prf.variance"] > 0)
        d1 = d1.select(d1["intensity.sum.variance"] > 0)

```

graeme-winter committed:
Remove reflections with -ve variance

Co-authored-by: jbeilstenedmands <30625594+jbeilstenedmands@users.noreply.github.com>

graeme-winter committed:
Merge branch 'master' into fp3

graeme-winter committed:
API change
